### PR TITLE
[Configuration] Authentication plugin

### PR DESCRIPTION
**Pull request by ityuhui**
[Authentication plugin](https://github.com/kubernetes-client/c/issues/4#issuecomment-602377670) for people to register new authentication methods dynamically. 

1. Authentication provider plugin framework

2. An instance of authentication provider plugin for OIDC (OpenID Connect)

### CODE REVIEW

File: kubernetes/config/authn_plugin/authn_plugin_util.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    if (!http_client) {
+        fprintf(stderr, "%s: Cannot create http client. [%s].\n", fname, strerror(errno));
+        return -1;
+    }
+    apiClient_invoke(http_client, NULL, NULL, NULL, NULL, NULL, contentType, post_data, type);
+
+    int rc = http_client->response_code;
+    switch (rc) {
+    case HTTP_RC_OK:
+        *p_http_response = strdup(http_client->dataReceived);
```
Comment (brendandburns) [inline kubernetes/config/authn_plugin/authn_plugin_util.c:18]: `strndup` for safety.

File: kubernetes/config/authn_plugin/authn_plugin_util.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#ifdef  __cplusplus
+extern "C" {
+#endif
+
+#define HTTP_REQUEST_GET "GET"
+#define HTTP_REQUEST_POST "POST"
+#define HTTP_REQUEST_DELETE "DELETE"
+#define HTTP_REQUEST_PATCH "PATCH"
+
+    typedef enum shc_http_rc_t {
```
Comment (brendandburns) [inline kubernetes/config/authn_plugin/authn_plugin_util.h:15]: fix indentation.

File: kubernetes/config/authn_plugin/authn_plugin_util.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ * apiKeys      : key-value list including basic or bearer token for HTTP authentication
+ * contentType  : string list including HTTP content type
+ * post_data    : post data string when request_type is "POST"
+ *
+ * OUT:
+ * p_http_response          : pointer that pointing to HTTP response raw string, the memory is allocated in shc_request, but user should free it out of shc_request after it is useless.
+ * p_http_response_length   : pointer that pointing to the length of p_http_response
+ *
+ */
+    int shc_request(char **p_http_response, int *p_http_response_length, char *request_type, const char *url, sslConfig_t * sc, list_t * apiKeys, list_t * contentType, char *post_data);
```
Comment (brendandburns) [inline kubernetes/config/authn_plugin/authn_plugin_util.h:46]: fix indent.

File: kubernetes/config/authn_plugin/authn_plugin_util.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * IN:
+ * json_string  : a json string that includes key-value string
+ * key          : the key needs to find
+ *
+ * OUT:
+ * None
+ *
+ */
+    char *shc_get_string_from_json(const char *json_string, const char *key);
```
Comment (brendandburns) [inline kubernetes/config/authn_plugin/authn_plugin_util.h:70]: fix indent.

File: kubernetes/config/authn_plugin/plugins/oidc/libkubernetes_oidc.c [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    if (!p) {
+        fprintf(stderr, "%s: The token <%s> is not a valid JWT token.\n", fname, token_string);
+        goto end;
+    }
+
+    int base64_padding_length = 4 - strlen(p) % 4;
+    char base64_padding_string[BASE64_PADDING_STRING_SIZE];
+    memset(base64_padding_string, 0, sizeof(base64_padding_string));
+    for (int i = 0; i < base64_padding_length; i++) {
+        strncat(base64_padding_string, BASE64_PADDING_CHAR, 1);
```
Comment (brendandburns) [inline kubernetes/config/authn_plugin/plugins/oidc/libkubernetes_oidc.c:59]: given that you just `memset` everything to zero, why do you need to strncat the padding?

Comment (ityuhui) [reply kubernetes/config/authn_plugin/plugins/oidc/libkubernetes_oidc.c:59]: This code snippets is used to generate base64 padding string for "base64_padding_string" ( the content is "=" , "==",  or "==="  )

for the first time (i = 1),
```strncat``` actually works same as ```strncpy```  
"" (empty string) --> “=”

but the second time (i = 2) or the third time (i=3)
```strncat``` will do the string concatenation.
"="  --> "=="
"=="  --> "==="

File: kubernetes/config/authn_plugin/authn_plugin_util.c (hunk 1 above)
Comment (ityuhui) [reply kubernetes/config/authn_plugin/authn_plugin_util.c]: done

File: kubernetes/config/authn_plugin/authn_plugin_util.h (hunk 2 above)
Comment (ityuhui) [reply kubernetes/config/authn_plugin/authn_plugin_util.h]: done

File: kubernetes/config/authn_plugin/authn_plugin_util.h (hunk 3 above)
Comment (ityuhui) [reply kubernetes/config/authn_plugin/authn_plugin_util.h]: done

File: kubernetes/config/authn_plugin/authn_plugin_util.h (hunk 4 above)
Comment (ityuhui) [reply kubernetes/config/authn_plugin/authn_plugin_util.h]: done

**Comment by ityuhui** [comment]
Hi @brendandburns 

The comments are addressed.

**Pull request merged**